File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-shared-5-4/GrossCashflow.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class GrossCashflow : ISerialized
    {
        public GrossCashflow(XmlNode xmlNode)
        {
            XmlNode cashflowIdNode = xmlNode.SelectSingleNode("cashflowId");
            
            if (cashflowIdNode != null)
            {
                if (cashflowIdNode.Attributes["href"] != null || cashflowIdNode.Attributes["id"] != null) 
                {
                    if (cashflowIdNode.Attributes["id"] != null) 
                    {
                        cashflowIdIDRef_ = cashflowIdNode.Attributes["id"].Value;
                        CashflowId ob = new CashflowId(cashflowIdNode);
                        IDManager.SetID(cashflowIdIDRef_, ob);
                    }
                    else if (cashflowIdNode.Attributes["href"] != null)
                    {
                        cashflowIdIDRef_ = cashflowIdNode.Attributes["href"].Value;
                    }
                    else
                    {
                        cashflowId_ = new CashflowId(cashflowIdNode);
                    }
                }
                else
                {
                    cashflowId_ = new CashflowId(cashflowIdNode);
                }
            }
            
        
            XmlNode partyTradeIdentifierReferenceNode = xmlNode.SelectSingleNode("partyTradeIdentifierReference");
            
            if (partyTradeIdentifierReferenceNode != null)
            {
                if (partyTradeIdentifierReferenceNode.Attributes["href"] != null || partyTradeIdentifierReferenceNode.Attributes["id"] != null) 
                {
                    if (partyTradeIdentifierReferenceNode.Attributes["id"] != null) 
                    {
                        partyTradeIdentifierReferenceIDRef_ = partyTradeIdentifierReferenceNode.Attributes["id"].Value;
                        PartyTradeIdentifierReference ob = new PartyTradeIdentifierReference(partyTradeIdentifierReferenceNode);
                        IDManager.SetID(partyTradeIdentifierReferenceIDRef_, ob);
                    }
                    else if (partyTradeIdentifierReferenceNode.Attributes["href"] != null)
                    {
                        partyTradeIdentifierReferenceIDRef_ = partyTradeIdentifierReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        partyTradeIdentifierReference_ = new PartyTradeIdentifierReference(partyTradeIdentifierReferenceNode);
                    }
                }
                else
                {
                    partyTradeIdentifierReference_ = new PartyTradeIdentifierReference(partyTradeIdentifierReferenceNode);
                }
            }
            
        
            XmlNode payerPartyReferenceNode = xmlNode.SelectSingleNode("payerPartyReference");
            
            if (payerPartyReferenceNode != null)
            {
                if (payerPartyReferenceNode.Attributes["href"] != null || payerPartyReferenceNode.Attributes["id"] != null) 
                {
                    if (payerPartyReferenceNode.Attributes["id"] != null) 
                    {
                        payerPartyReferenceIDRef_ = payerPartyReferenceNode.Attributes["id"].Value;
                        PartyReference ob = new PartyReference(payerPartyReferenceNode);
                        IDManager.SetID(payerPartyReferenceIDRef_, ob);
                    }
                    else if (payerPartyReferenceNode.Attributes["href"] != null)
                    {
                        payerPartyReferenceIDRef_ = payerPartyReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        payerPartyReference_ = new PartyReference(payerPartyReferenceNode);
                    }
                }
                else
                {
                    payerPartyReference_ = new PartyReference(payerPartyReferenceNode);
                }
            }
            
        
            XmlNode payerAccountReferenceNode = xmlNode.SelectSingleNode("payerAccountReference");
            
            if (payerAccountReferenceNode != null)
            {
                if (payerAccountReferenceNode.Attributes["href"] != null || payerAccountReferenceNode.Attributes["id"] != null) 
                {
                    if (payerAccountReferenceNode.Attributes["id"] != null) 
                    {
                        payerAccountReferenceIDRef_ = payerAccountReferenceNode.Attributes["id"].Value;
                        AccountReference ob = new AccountReference(payerAccountReferenceNode);
                        IDManager.SetID(payerAccountReferenceIDRef_, ob);
                    }
                    else if (payerAccountReferenceNode.Attributes["href"] != null)
                    {
                        payerAccountReferenceIDRef_ = payerAccountReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        payerAccountReference_ = new AccountReference(payerAccountReferenceNode);
                    }
                }
                else
                {
                    payerAccountReference_ = new AccountReference(payerAccountReferenceNode);
                }
            }
            
        
            XmlNode receiverPartyReferenceNode = xmlNode.SelectSingleNode("receiverPartyReference");
            
            if (receiverPartyReferenceNode != null)
            {
                if (receiverPartyReferenceNode.Attributes["href"] != null || receiverPartyReferenceNode.Attributes["id"] != null) 
                {
                    if (receiverPartyReferenceNode.Attributes["id"] != null) 
                    {
                        receiverPartyReferenceIDRef_ = receiverPartyReferenceNode.Attributes["id"].Value;
                        PartyReference ob = new PartyReference(receiverPartyReferenceNode);
                        IDManager.SetID(receiverPartyReferenceIDRef_, ob);
                    }
                    else if (receiverPartyReferenceNode.Attributes["href"] != null)
                    {
                        receiverPartyReferenceIDRef_ = receiverPartyReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        receiverPartyReference_ = new PartyReference(receiverPartyReferenceNode);
                    }
                }
                else
                {
                    receiverPartyReference_ = new PartyReference(receiverPartyReferenceNode);
                }
            }
            
        
            XmlNode receiverAccountReferenceNode = xmlNode.SelectSingleNode("receiverAccountReference");
            
            if (receiverAccountReferenceNode != null)
            {
                if (receiverAccountReferenceNode.Attributes["href"] != null || receiverAccountReferenceNode.Attributes["id"] != null) 
                {
                    if (receiverAccountReferenceNode.Attributes["id"] != null) 
                    {
                        receiverAccountReferenceIDRef_ = receiverAccountReferenceNode.Attributes["id"].Value;
                        AccountReference ob = new AccountReference(receiverAccountReferenceNode);
                        IDManager.SetID(receiverAccountReferenceIDRef_, ob);
                    }
                    else if (receiverAccountReferenceNode.Attributes["href"] != null)
                    {
                        receiverAccountReferenceIDRef_ = receiverAccountReferenceNode.Attributes["href"].Value;
                    }
                    else
                    {
                        receiverAccountReference_ = new AccountReference(receiverAccountReferenceNode);
                    }
                }
                else
                {
                    receiverAccountReference_ = new AccountReference(receiverAccountReferenceNode);
                }
            }
            
        
            XmlNode cashflowAmountNode = xmlNode.SelectSingleNode("cashflowAmount");
            
            if (cashflowAmountNode != null)
            {
                if (cashflowAmountNode.Attributes["href"] != null || cashflowAmountNode.Attributes["id"] != null) 
                {
                    if (cashflowAmountNode.Attributes["id"] != null) 
                    {
                        cashflowAmountIDRef_ = cashflowAmountNode.Attributes["id"].Value;
                        Money ob = new Money(cashflowAmountNode);
                        IDManager.SetID(cashflowAmountIDRef_, ob);
                    }
                    else if (cashflowAmountNode.Attributes["href"] != null)
                    {
                        cashflowAmountIDRef_ = cashflowAmountNode.Attributes["href"].Value;
                    }
                    else
                    {
                        cashflowAmount_ = new Money(cashflowAmountNode);
                    }
                }
                else
                {
                    cashflowAmount_ = new Money(cashflowAmountNode);
                }
            }
            
        
            XmlNode cashflowTypeNode = xmlNode.SelectSingleNode("cashflowType");
            
            if (cashflowTypeNode != null)
            {
                if (cashflowTypeNode.Attributes["href"] != null || cashflowTypeNode.Attributes["id"] != null) 
                {
                    if (cashflowTypeNode.Attributes["id"] != null) 
                    {
                        cashflowTypeIDRef_ = cashflowTypeNode.Attributes["id"].Value;
                        CashflowType ob = new CashflowType(cashflowTypeNode);
                        IDManager.SetID(cashflowTypeIDRef_, ob);
                    }
                    else if (cashflowTypeNode.Attributes["href"] != null)
                    {
                        cashflowTypeIDRef_ = cashflowTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        cashflowType_ = new CashflowType(cashflowTypeNode);
                    }
                }
                else
                {
                    cashflowType_ = new CashflowType(cashflowTypeNode);
                }
            }
            
        
        }
        
    
        #region CashflowId_
        private CashflowId cashflowId_;
        public CashflowId CashflowId_
        {
            get
            {
                if (this.cashflowId_ != null)
                {
                    return this.cashflowId_; 
                }
                else if (this.cashflowIdIDRef_ != null)
                {
                    cashflowId_ = IDManager.getID(cashflowIdIDRef_) as CashflowId;
                    return this.cashflowId_; 
                }
                else
                {
                      return this.cashflowId_; 
                }
            }
            set
            {
                if (this.cashflowId_ != value)
                {
                    this.cashflowId_ = value;
                }
            }
        }
        #endregion
        
        public string cashflowIdIDRef_ { get; set; }
        #region PartyTradeIdentifierReference_
        private PartyTradeIdentifierReference partyTradeIdentifierReference_;
        public PartyTradeIdentifierReference PartyTradeIdentifierReference_
        {
            get
            {
                if (this.partyTradeIdentifierReference_ != null)
                {
                    return this.partyTradeIdentifierReference_; 
                }
                else if (this.partyTradeIdentifierReferenceIDRef_ != null)
                {
                    partyTradeIdentifierReference_ = IDManager.getID(partyTradeIdentifierReferenceIDRef_) as PartyTradeIdentifierReference;
                    return this.partyTradeIdentifierReference_; 
                }
                else
                {
                      return this.partyTradeIdentifierReference_; 
                }
            }
            set
            {
                if (this.partyTradeIdentifierReference_ != value)
                {
                    this.partyTradeIdentifierReference_ = value;
                }
            }
        }
        #endregion
        
        public string partyTradeIdentifierReferenceIDRef_ { get; set; }
        #region PayerPartyReference_
        private PartyReference payerPartyReference_;
        public PartyReference PayerPartyReference_
        {
            get
            {
                if (this.payerPartyReference_ != null)
                {
                    return this.payerPartyReference_; 
                }
                else if (this.payerPartyReferenceIDRef_ != null)
                {
                    payerPartyReference_ = IDManager.getID(payerPartyReferenceIDRef_) as PartyReference;
                    return this.payerPartyReference_; 
                }
                else
                {
                      return this.payerPartyReference_; 
                }
            }
            set
            {
                if (this.payerPartyReference_ != value)
                {
                    this.payerPartyReference_ = value;
                }
            }
        }
        #endregion
        
        public string payerPartyReferenceIDRef_ { get; set; }
        #region PayerAccountReference_
        private AccountReference payerAccountReference_;
        public AccountReference PayerAccountReference_
        {
            get
            {
                if (this.payerAccountReference_ != null)
                {
                    return this.payerAccountReference_; 
                }
                else if (this.payerAccountReferenceIDRef_ != null)
                {
                    payerAccountReference_ = IDManager.getID(payerAccountReferenceIDRef_) as AccountReference;
                    return this.payerAccountReference_; 
                }
                else
                {
                      return this.payerAccountReference_; 
                }
            }
            set
            {
                if (this.payerAccountReference_ != value)
                {
                    this.payerAccountReference_ = value;
                }
            }
        }
        #endregion
        
        public string payerAccountReferenceIDRef_ { get; set; }
        #region ReceiverPartyReference_
        private PartyReference receiverPartyReference_;
        public PartyReference ReceiverPartyReference_
        {
            get
            {
                if (this.receiverPartyReference_ != null)
                {
                    return this.receiverPartyReference_; 
                }
                else if (this.receiverPartyReferenceIDRef_ != null)
                {
                    receiverPartyReference_ = IDManager.getID(receiverPartyReferenceIDRef_) as PartyReference;
                    return this.receiverPartyReference_; 
                }
                else
                {
                      return this.receiverPartyReference_; 
                }
            }
            set
            {
                if (this.receiverPartyReference_ != value)
                {
                    this.receiverPartyReference_ = value;
                }
            }
        }
        #endregion
        
        public string receiverPartyReferenceIDRef_ { get; set; }
        #region ReceiverAccountReference_
        private AccountReference receiverAccountReference_;
        public AccountReference ReceiverAccountReference_
        {
            get
            {
                if (this.receiverAccountReference_ != null)
                {
                    return this.receiverAccountReference_; 
                }
                else if (this.receiverAccountReferenceIDRef_ != null)
                {
                    receiverAccountReference_ = IDManager.getID(receiverAccountReferenceIDRef_) as AccountReference;
                    return this.receiverAccountReference_; 
                }
                else
                {
                      return this.receiverAccountReference_; 
                }
            }
            set
            {
                if (this.receiverAccountReference_ != value)
                {
                    this.receiverAccountReference_ = value;
                }
            }
        }
        #endregion
        
        public string receiverAccountReferenceIDRef_ { get; set; }
        #region CashflowAmount_
        private Money cashflowAmount_;
        public Money CashflowAmount_
        {
            get
            {
                if (this.cashflowAmount_ != null)
                {
                    return this.cashflowAmount_; 
                }
                else if (this.cashflowAmountIDRef_ != null)
                {
                    cashflowAmount_ = IDManager.getID(cashflowAmountIDRef_) as Money;
                    return this.cashflowAmount_; 
                }
                else
                {
                      return this.cashflowAmount_; 
                }
            }
            set
            {
                if (this.cashflowAmount_ != value)
                {
                    this.cashflowAmount_ = value;
                }
            }
        }
        #endregion
        
        public string cashflowAmountIDRef_ { get; set; }
        #region CashflowType_
        private CashflowType cashflowType_;
        public CashflowType CashflowType_
        {
            get
            {
                if (this.cashflowType_ != null)
                {
                    return this.cashflowType_; 
                }
                else if (this.cashflowTypeIDRef_ != null)
                {
                    cashflowType_ = IDManager.getID(cashflowTypeIDRef_) as CashflowType;
                    return this.cashflowType_; 
                }
                else
                {
                      return this.cashflowType_; 
                }
            }
            set
            {
                if (this.cashflowType_ != value)
                {
                    this.cashflowType_ = value;
                }
            }
        }
        #endregion
        
        public string cashflowTypeIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-shared-5-4/GrossCashflow.cpp
// GrossCashflow.cpp 
#include "GrossCashflow.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

GrossCashflow::GrossCashflow(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //cashflowIdNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* cashflowIdNode = xmlNode->FirstChildElement("cashflowId");

   if(cashflowIdNode){cashflowIdIsNull_ = false;}
   else{cashflowIdIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- cashflowIdNode , address : " << cashflowIdNode << std::endl;
   #endif
   if(cashflowIdNode)
   {
      if(cashflowIdNode->Attribute("href") || cashflowIdNode->Attribute("id"))
      {
          if(cashflowIdNode->Attribute("id"))
          {
             cashflowIdIDRef_ = cashflowIdNode->Attribute("id");
             cashflowId_ = boost::shared_ptr<CashflowId>(new CashflowId(cashflowIdNode));
             cashflowId_->setID(cashflowIdIDRef_);
             IDManager::instance().setID(cashflowIdIDRef_,cashflowId_);
          }
          else if(cashflowIdNode->Attribute("href")) { cashflowIdIDRef_ = cashflowIdNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { cashflowId_ = boost::shared_ptr<CashflowId>(new CashflowId(cashflowIdNode));}
   }

   //partyTradeIdentifierReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* partyTradeIdentifierReferenceNode = xmlNode->FirstChildElement("partyTradeIdentifierReference");

   if(partyTradeIdentifierReferenceNode){partyTradeIdentifierReferenceIsNull_ = false;}
   else{partyTradeIdentifierReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- partyTradeIdentifierReferenceNode , address : " << partyTradeIdentifierReferenceNode << std::endl;
   #endif
   if(partyTradeIdentifierReferenceNode)
   {
      if(partyTradeIdentifierReferenceNode->Attribute("href") || partyTradeIdentifierReferenceNode->Attribute("id"))
      {
          if(partyTradeIdentifierReferenceNode->Attribute("id"))
          {
             partyTradeIdentifierReferenceIDRef_ = partyTradeIdentifierReferenceNode->Attribute("id");
             partyTradeIdentifierReference_ = boost::shared_ptr<PartyTradeIdentifierReference>(new PartyTradeIdentifierReference(partyTradeIdentifierReferenceNode));
             partyTradeIdentifierReference_->setID(partyTradeIdentifierReferenceIDRef_);
             IDManager::instance().setID(partyTradeIdentifierReferenceIDRef_,partyTradeIdentifierReference_);
          }
          else if(partyTradeIdentifierReferenceNode->Attribute("href")) { partyTradeIdentifierReferenceIDRef_ = partyTradeIdentifierReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { partyTradeIdentifierReference_ = boost::shared_ptr<PartyTradeIdentifierReference>(new PartyTradeIdentifierReference(partyTradeIdentifierReferenceNode));}
   }

   //payerPartyReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* payerPartyReferenceNode = xmlNode->FirstChildElement("payerPartyReference");

   if(payerPartyReferenceNode){payerPartyReferenceIsNull_ = false;}
   else{payerPartyReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- payerPartyReferenceNode , address : " << payerPartyReferenceNode << std::endl;
   #endif
   if(payerPartyReferenceNode)
   {
      if(payerPartyReferenceNode->Attribute("href") || payerPartyReferenceNode->Attribute("id"))
      {
          if(payerPartyReferenceNode->Attribute("id"))
          {
             payerPartyReferenceIDRef_ = payerPartyReferenceNode->Attribute("id");
             payerPartyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(payerPartyReferenceNode));
             payerPartyReference_->setID(payerPartyReferenceIDRef_);
             IDManager::instance().setID(payerPartyReferenceIDRef_,payerPartyReference_);
          }
          else if(payerPartyReferenceNode->Attribute("href")) { payerPartyReferenceIDRef_ = payerPartyReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { payerPartyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(payerPartyReferenceNode));}
   }

   //payerAccountReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* payerAccountReferenceNode = xmlNode->FirstChildElement("payerAccountReference");

   if(payerAccountReferenceNode){payerAccountReferenceIsNull_ = false;}
   else{payerAccountReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- payerAccountReferenceNode , address : " << payerAccountReferenceNode << std::endl;
   #endif
   if(payerAccountReferenceNode)
   {
      if(payerAccountReferenceNode->Attribute("href") || payerAccountReferenceNode->Attribute("id"))
      {
          if(payerAccountReferenceNode->Attribute("id"))
          {
             payerAccountReferenceIDRef_ = payerAccountReferenceNode->Attribute("id");
             payerAccountReference_ = boost::shared_ptr<AccountReference>(new AccountReference(payerAccountReferenceNode));
             payerAccountReference_->setID(payerAccountReferenceIDRef_);
             IDManager::instance().setID(payerAccountReferenceIDRef_,payerAccountReference_);
          }
          else if(payerAccountReferenceNode->Attribute("href")) { payerAccountReferenceIDRef_ = payerAccountReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { payerAccountReference_ = boost::shared_ptr<AccountReference>(new AccountReference(payerAccountReferenceNode));}
   }

   //receiverPartyReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* receiverPartyReferenceNode = xmlNode->FirstChildElement("receiverPartyReference");

   if(receiverPartyReferenceNode){receiverPartyReferenceIsNull_ = false;}
   else{receiverPartyReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- receiverPartyReferenceNode , address : " << receiverPartyReferenceNode << std::endl;
   #endif
   if(receiverPartyReferenceNode)
   {
      if(receiverPartyReferenceNode->Attribute("href") || receiverPartyReferenceNode->Attribute("id"))
      {
          if(receiverPartyReferenceNode->Attribute("id"))
          {
             receiverPartyReferenceIDRef_ = receiverPartyReferenceNode->Attribute("id");
             receiverPartyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(receiverPartyReferenceNode));
             receiverPartyReference_->setID(receiverPartyReferenceIDRef_);
             IDManager::instance().setID(receiverPartyReferenceIDRef_,receiverPartyReference_);
          }
          else if(receiverPartyReferenceNode->Attribute("href")) { receiverPartyReferenceIDRef_ = receiverPartyReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { receiverPartyReference_ = boost::shared_ptr<PartyReference>(new PartyReference(receiverPartyReferenceNode));}
   }

   //receiverAccountReferenceNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* receiverAccountReferenceNode = xmlNode->FirstChildElement("receiverAccountReference");

   if(receiverAccountReferenceNode){receiverAccountReferenceIsNull_ = false;}
   else{receiverAccountReferenceIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- receiverAccountReferenceNode , address : " << receiverAccountReferenceNode << std::endl;
   #endif
   if(receiverAccountReferenceNode)
   {
      if(receiverAccountReferenceNode->Attribute("href") || receiverAccountReferenceNode->Attribute("id"))
      {
          if(receiverAccountReferenceNode->Attribute("id"))
          {
             receiverAccountReferenceIDRef_ = receiverAccountReferenceNode->Attribute("id");
             receiverAccountReference_ = boost::shared_ptr<AccountReference>(new AccountReference(receiverAccountReferenceNode));
             receiverAccountReference_->setID(receiverAccountReferenceIDRef_);
             IDManager::instance().setID(receiverAccountReferenceIDRef_,receiverAccountReference_);
          }
          else if(receiverAccountReferenceNode->Attribute("href")) { receiverAccountReferenceIDRef_ = receiverAccountReferenceNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { receiverAccountReference_ = boost::shared_ptr<AccountReference>(new AccountReference(receiverAccountReferenceNode));}
   }

   //cashflowAmountNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* cashflowAmountNode = xmlNode->FirstChildElement("cashflowAmount");

   if(cashflowAmountNode){cashflowAmountIsNull_ = false;}
   else{cashflowAmountIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- cashflowAmountNode , address : " << cashflowAmountNode << std::endl;
   #endif
   if(cashflowAmountNode)
   {
      if(cashflowAmountNode->Attribute("href") || cashflowAmountNode->Attribute("id"))
      {
          if(cashflowAmountNode->Attribute("id"))
          {
             cashflowAmountIDRef_ = cashflowAmountNode->Attribute("id");
             cashflowAmount_ = boost::shared_ptr<Money>(new Money(cashflowAmountNode));
             cashflowAmount_->setID(cashflowAmountIDRef_);
             IDManager::instance().setID(cashflowAmountIDRef_,cashflowAmount_);
          }
          else if(cashflowAmountNode->Attribute("href")) { cashflowAmountIDRef_ = cashflowAmountNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { cashflowAmount_ = boost::shared_ptr<Money>(new Money(cashflowAmountNode));}
   }

   //cashflowTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* cashflowTypeNode = xmlNode->FirstChildElement("cashflowType");

   if(cashflowTypeNode){cashflowTypeIsNull_ = false;}
   else{cashflowTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- cashflowTypeNode , address : " << cashflowTypeNode << std::endl;
   #endif
   if(cashflowTypeNode)
   {
      if(cashflowTypeNode->Attribute("href") || cashflowTypeNode->Attribute("id"))
      {
          if(cashflowTypeNode->Attribute("id"))
          {
             cashflowTypeIDRef_ = cashflowTypeNode->Attribute("id");
             cashflowType_ = boost::shared_ptr<CashflowType>(new CashflowType(cashflowTypeNode));
             cashflowType_->setID(cashflowTypeIDRef_);
             IDManager::instance().setID(cashflowTypeIDRef_,cashflowType_);
          }
          else if(cashflowTypeNode->Attribute("href")) { cashflowTypeIDRef_ = cashflowTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { cashflowType_ = boost::shared_ptr<CashflowType>(new CashflowType(cashflowTypeNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<CashflowId> GrossCashflow::getCashflowId()
{
   if(!this->cashflowIdIsNull_){
        if(cashflowIdIDRef_ != ""){
             return boost::shared_static_cast<CashflowId>(IDManager::instance().getID(cashflowIdIDRef_));
        }else{
             return this->cashflowId_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CashflowId>();
   }
}
boost::shared_ptr<PartyTradeIdentifierReference> GrossCashflow::getPartyTradeIdentifierReference()
{
   if(!this->partyTradeIdentifierReferenceIsNull_){
        if(partyTradeIdentifierReferenceIDRef_ != ""){
             return boost::shared_static_cast<PartyTradeIdentifierReference>(IDManager::instance().getID(partyTradeIdentifierReferenceIDRef_));
        }else{
             return this->partyTradeIdentifierReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyTradeIdentifierReference>();
   }
}
boost::shared_ptr<PartyReference> GrossCashflow::getPayerPartyReference()
{
   if(!this->payerPartyReferenceIsNull_){
        if(payerPartyReferenceIDRef_ != ""){
             return boost::shared_static_cast<PartyReference>(IDManager::instance().getID(payerPartyReferenceIDRef_));
        }else{
             return this->payerPartyReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyReference>();
   }
}
boost::shared_ptr<AccountReference> GrossCashflow::getPayerAccountReference()
{
   if(!this->payerAccountReferenceIsNull_){
        if(payerAccountReferenceIDRef_ != ""){
             return boost::shared_static_cast<AccountReference>(IDManager::instance().getID(payerAccountReferenceIDRef_));
        }else{
             return this->payerAccountReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AccountReference>();
   }
}
boost::shared_ptr<PartyReference> GrossCashflow::getReceiverPartyReference()
{
   if(!this->receiverPartyReferenceIsNull_){
        if(receiverPartyReferenceIDRef_ != ""){
             return boost::shared_static_cast<PartyReference>(IDManager::instance().getID(receiverPartyReferenceIDRef_));
        }else{
             return this->receiverPartyReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<PartyReference>();
   }
}
boost::shared_ptr<AccountReference> GrossCashflow::getReceiverAccountReference()
{
   if(!this->receiverAccountReferenceIsNull_){
        if(receiverAccountReferenceIDRef_ != ""){
             return boost::shared_static_cast<AccountReference>(IDManager::instance().getID(receiverAccountReferenceIDRef_));
        }else{
             return this->receiverAccountReference_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AccountReference>();
   }
}
boost::shared_ptr<Money> GrossCashflow::getCashflowAmount()
{
   if(!this->cashflowAmountIsNull_){
        if(cashflowAmountIDRef_ != ""){
             return boost::shared_static_cast<Money>(IDManager::instance().getID(cashflowAmountIDRef_));
        }else{
             return this->cashflowAmount_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Money>();
   }
}
boost::shared_ptr<CashflowType> GrossCashflow::getCashflowType()
{
   if(!this->cashflowTypeIsNull_){
        if(cashflowTypeIDRef_ != ""){
             return boost::shared_static_cast<CashflowType>(IDManager::instance().getID(cashflowTypeIDRef_));
        }else{
             return this->cashflowType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<CashflowType>();
   }
}
}

